### PR TITLE
Remove support for creating Continuous Aggregates with old format

### PR DESCRIPTION
**Pull request by fabriziomello**
Timescale 2.7 released a new version of Continuous Aggregate (#4269) that store the final aggregation state instead of the byte array of the partial aggregate state, offering multiple opportunities of optimizations as well a more compact form.

In 2.10.0 released on February 2023 the Continuous Aggregate old format deprecation was announced.

With this PR the ability of creating Continuous Aggregate in the old format was removed, but we still support migrate from the old to the new format by running the `cagg_migrate` procedure.

This is the continuation of the PR #5977 started by @pdipesh02.

References:
https://docs.timescale.com/api/latest/continuous-aggregates/cagg_migrate/ https://github.com/timescale/timescaledb/releases/tag/2.10.0 https://github.com/timescale/timescaledb/releases/tag/2.7.0 https://github.com/timescale/timescaledb/pull/5977

### CODE REVIEW

File: tsl/src/continuous_aggs/repair.c [hunk 1]
```diff
@@ -23,12 +23,20 @@ cagg_rebuild_view_definition(ContinuousAgg *agg, Hypertable *mat_ht, bool force_
 	int sec_ctx;
 	Oid uid, saved_uid;
 
+	bool finalized = ContinuousAggIsFinalized(agg);
+	if (!finalized)
+	{
+		ereport(WARNING,
+				(errmsg("Continuous Aggregates with partials is not supported anymore."),
```
Comment (jnidzwetzki) [inline tsl/src/continuous_aggs/repair.c:30]:
```suggestion
				(errmsg("Continuous Aggregates with partials are not supported anymore."),
```

File: tsl/src/continuous_aggs/common.c [hunk 2]
```diff
@@ -432,13 +387,11 @@ cagg_agg_validate(Node *node, void *context)
...
 {
-/*
- * For now deprecate partial aggregates on release builds only.
- * Once migration tests are made compatible with PG15 enable deprecation
- * on debug builds as well.
- */
-#ifndef DEBUG
-#if PG15_GE
+	/*
+	 * For now deprecate partial aggregates on release builds only.
```
Comment (jnidzwetzki) [inline tsl/src/continuous_aggs/common.c:391]: I guess this comment can be removed now.

File: .unreleased/pr_6360 [hunk 3]
```diff
@@ -0,0 +1,3 @@
+Implements: #6360 Remove old Continuous Aggregate format
```
Comment (jnidzwetzki) [inline .unreleased/pr_6360:1]: CAggs with old format are still supported afterward. However, no new ones can be created.

```suggestion
Implements: #6360 Remove support for creating Continuous Aggregates with old format 
```

File: sql/updates/post-update.sql [hunk 4]
```diff
@@ -1,67 +1,107 @@
...
     EXCEPTION WHEN OTHERS THEN RAISE;
 END
-$$;
+$$ LANGUAGE PLPGSQL;
+CALL _timescaledb_internal.fixup_caggs();
+DROP PROCEDURE _timescaledb_internal.fixup_caggs();
 
 -- For tsdb >= v2.10.0 apply the cagg repair when necessary
-DO $$
+CREATE OR REPLACE PROCEDURE _timescaledb_internal.fixup_caggs() AS
```
Comment (jnidzwetzki) [inline sql/updates/post-update.sql:63]: Could you elaborate on why the function is re-created at this point? `_timescaledb_internal.fixup_caggs` was already created in line 2.

File: sql/updates/post-update.sql [hunk 5]
```diff
@@ -1,67 +1,107 @@
...
+            END IF;
+        END LOOP;
+    ELSIF ts_major::INTEGER = 2 AND ts_minor::INTEGER >= 7 THEN
+        FOR vname, mat_ht_id, materialized_only, finalized IN select format('%I.%I', cagg.user_view_schema,
+            cagg.user_view_name)::regclass, cagg.mat_hypertable_id, cagg.materialized_only, cagg.finalized from _timescaledb_catalog.continuous_agg cagg
+        LOOP
+            IF ts_minor::INTEGER < 12 THEN
+                SET log_error_verbosity TO VERBOSE;
+                CALL _timescaledb_internal.post_update_cagg_try_repair(vname, false);
+            ELSIF ts_minor::INTEGER = 12 AND finalized IS FALSE THEN
```
Comment (jnidzwetzki) [inline sql/updates/post-update.sql:41]: I don't understand the reason for the following lines at the moment. Could you add a comment on why we call  `public.cagg_migrate` for this specific version?

**Review by jnidzwetzki — COMMENTED** [review]
Overall, it looks good to me. However, I have a few questions regarding the changes in the `fixup_caggs` function before approval.

File: .unreleased/pr_6360 (hunk 3 above)
Comment (fabriziomello) [reply .unreleased/pr_6360:1]: Fixed

File: tsl/src/continuous_aggs/common.c (hunk 2 above)
Comment (fabriziomello) [reply tsl/src/continuous_aggs/common.c:391]: Removed!

File: sql/updates/post-update.sql (hunk 4 above)
Comment (fabriziomello) [reply sql/updates/post-update.sql:63]: TBH I really don't remember why Dipesh did it in that way... need to deep dive more on it. For sure there are a good reason.

File: tsl/src/continuous_aggs/repair.c [hunk 6]
```diff
@@ -23,12 +23,20 @@ cagg_rebuild_view_definition(ContinuousAgg *agg, Hypertable *mat_ht, bool force_
...
+				 errdetail("Migrate the Continuous Aggregates to finalized form to rebuild.")));
```
[the remark this answers is not in the document]
Comment (fabriziomello) [reply tsl/src/continuous_aggs/repair.c:31]: Good idea. Added!!!

File: sql/updates/post-update.sql [hunk 7]
```diff
@@ -1,67 +1,107 @@
 -- needed post 1.7.0 to fixup continuous aggregates created in 1.7.0 ---
-DO $$
+CREATE OR REPLACE PROCEDURE _timescaledb_internal.fixup_caggs() AS
+$$
```
Comment (svenklemm) [inline sql/updates/post-update.sql:3]: i think we shouldnt create functions in _timescaledb_internal even if only temporary

Comment (fabriziomello) [reply sql/updates/post-update.sql:3]: I know thanks... will change it!!!

File: sql/updates/post-update.sql (hunk 4 above)
Comment (fabriziomello) [reply sql/updates/post-update.sql]: Fixed!

File: sql/updates/post-update.sql (hunk 5 above)
Comment (fabriziomello) [reply sql/updates/post-update.sql]: Fixed!

File: sql/updates/post-update.sql (hunk 7 above)
Comment (fabriziomello) [reply sql/updates/post-update.sql]: Fixed!

**Review by svenklemm — APPROVED** [review]

File: tsl/src/continuous_aggs/repair.c [hunk 8]
```diff
@@ -17,18 +17,30 @@ static void
...
+	char *schema = NameStr(agg->data.user_view_schema);
 	ListCell *lc1, *lc2;
 	int sec_ctx;
 	Oid uid, saved_uid;
 
+	bool finalized = ContinuousAggIsFinalized(agg);
+	if (!finalized)
+	{
+		ereport(WARNING,
+				(errmsg("Continuous Aggregates with partials are not supported anymore."),
```
Comment (jnidzwetzki) [inline tsl/src/continuous_aggs/repair.c:30]: Nit:

```suggestion
				(errmsg("repairing Continuous Aggregates with partials are not supported anymore."),
```

**Review by jnidzwetzki — APPROVED** [review]
Added one minor comment. In addition, could you adjust the commit/pr message before merging? I would recommend to mention, that we only restrict the creation of these CAggs with this PR (e.g., `Remove support for creating non finalized CAggs`).

**Comment by fabriziomello** [comment]
> With this PR the ability of creating Continuous Aggregate in the old format was removed, but we still support migrate from the old to the new format by running the `cagg_migrate` procedure.

But I'm already mentioning it in the PR and commit messages:
`With this PR the ability of creating Continuous Aggregate in the old format was removed, but we still support migrate from the old to the new format by running the cagg_migrate procedure.`

**Comment by jnidzwetzki** [comment]
@fabriziomello You are right and my comment was not precise. Maybe we can change the commit title from `Remove old Continuous Aggregates format` to `Remove support for creating non finalized CAggs`) to reflect these changes better. But it's up to you.

**Comment by fabriziomello** [comment]
> @fabriziomello You are right and my comment was not precise. Maybe we can change the commit title from `Remove old Continuous Aggregates format` to `Remove support for creating non finalized CAggs` to reflect these changes better. But it's up to you.

What about `Remove support for creating Continuous Aggregates with old format` the same title we used for the changelog?

**Comment by jnidzwetzki** [comment]
@fabriziomello I like it, but it could be too long for the title of the commit. Therefore, I proposed something with less than 50 chars.

**Comment by fabriziomello** [comment]
> @fabriziomello I like it, but it could be too long for the title of the commit. Therefore, I proposed something with less than 50 chars.

I've made it shorter by just `s/Continuous Aggregates/CAggs/g`

**Pull request merged**